.. _alignment-detailed-section:

Alignment file evaluation & cleanup (:code:`BAM`)
=================================================

You are analyzing your RNA-Seq reads using a standard RNA-Seq pipeline. After performing quality control, you have utilized popular spliced aligners such as :code:`HISAT2` and :code:`STAR` to align the reads to the genome. While inspecting the alignments in IGV, you have noticed that some reads are spliced and aligned across different gene loci or intergenic regions. This raises concerns about the accuracy of these spliced alignments and whether the reads truly originate from these locations. Additionally, you are curious **if there is a systematic approach to validate these splice alignments**. Unfortunately, there are currently no available tools that can assess the results generated by the aligners. Consequently,

|

**we propose running splam as a new step in standard RNA-Seq analysis pipeline.**

|


splam takes a **sorted alignment file**, extracts all splice junctions, and scores all of them. Furthermore, splam cleans up the alignment file by removing spliced alignments with poor quality splice junctions, and output a new alignment file. 

splam solves of problem and gives us solutions of:

1. Which spliced alignments are good and which of them are in low confidence. 
2. Systematic way assess all splice alignments in an alignment file. Furthermore, it removes some spurious splice alignments, and improves the downstream transcriptome assembly. 
3. Potentially, splam can help us undertand what are the biases of splice aligners.
4. There are more potential usage to be explored!

Before diving into details of each step, this is the overview figure of the alignment file cleanup ideogram:


.. _alignment-prepareintput:

Step 1: Preparing your input files
+++++++++++++++++++++++++++++++++++

The first step is to prepare three files for splam analysis:

1. An alignment file in :code:`BAM` format [`example file: SRR1352129_chr9_sub.bam <https://github.com/Kuanhao-Chao/splam/blob/main/test/SRR1352129_chr9_sub.bam>`_].  
2. A reference genome in :code:`FASTA` format [`example file: chr9_subset.fa <https://github.com/Kuanhao-Chao/splam/blob/main/test/chr9_subset.fa>`_].
3. The splam model, which you can find it here: `splam.pt <https://github.com/Kuanhao-Chao/splam/blob/main/model/splam_script.pt>`_

|

Step 2: Extracting splice junctions in your alignment file
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

.. code-block:: bash

    splam extract -P SRR1352129_chr9_sub.bam

|

Step 3: Scoring extracted splice junctions
++++++++++++++++++++++++++++++++++++++++++++++++++++++++

.. code-block:: bash

    splam score -G chr9_subset.fa -m ../model/splam_script.pt -o tmp_out tmp_out/junction.bed

|


Step 4: Cleaning up your alignment file
++++++++++++++++++++++++++++++++++++++++++++++++++++++++

.. code-block:: bash

    splam clean -o tmp_out -@ 5

|

Step 5: Visualization & reports
+++++++++++++++++++++++++++++++++++

.. code-block:: bash

    splam extract -P SRR1352129_chr9_sub.bam

|
|
